crates/engine/tree/src/tree/root.rs

```diff
@@ -1136,37 +1295,72 @@ where
-
-    // first collect all new accounts (not previously fetched)
-    for &hashed_address in state_update.accounts.keys() {
-        if !fetched_proof_targets.contains_key(&hashed_address) {
-            targets.insert(hashed_address, HashSet::default());
+/// Iterator that chunks proof targets.
+///
+/// This iterator will yield items of type [`Vec<MultiProofTargets>`], with each
+/// [`MultiProofTargets`] mapping having a maximum length of `accounts_chunk_size`, and each mapping
+/// value [`B256Set`] having a maximum length of `storages_chunk_size`.
+struct ChunkedProofTargets {
+    /// Sorted proof targets.
+    ///
+    /// We keep this sorted, because multiproofs are more efficient when proof targets are located
+    /// close to each other.
+    proof_targets: VecDeque<(B256, B256Set)>,
+    /// Number of accounts per chunk.
+    accounts_chunk_size: usize,
+    /// Number of storage slots per account per chunk.
+    storages_chunk_size: usize,
+}
+
+impl ChunkedProofTargets {
+    fn new(
+        proof_targets: MultiProofTargets,
+        accounts_chunk_size: usize,
+        storages_chunk_size: usize,
+    ) -> Self {
+        Self {
+            proof_targets: proof_targets
+                .into_iter()
+                .sorted_by_key(|(address, _)| *address)
+                .collect(),
+            accounts_chunk_size,
+            storages_chunk_size,
         }
     }
+}
+
+impl Iterator for ChunkedProofTargets {
+    type Item = Vec<MultiProofTargets>;
 
-    // then process storage slots for all accounts in the state update
-    for (hashed_address, storage) in &state_update.storages {
-        let fetched = fetched_proof_targets.get(hashed_address);
-        let mut changed_slots = storage
-            .storage
-            .keys()
-            .filter(|slot| !fetched.is_some_and(|f| f.contains(*slot)))
-            .peekable();
-
-        if changed_slots.peek().is_some() {
-            targets.entry(*hashed_address).or_default().extend(changed_slots);
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.proof_targets.is_empty() {
+            return None;
         }
-    }
 
-    targets
+        // Every address will be added to at least first chunk
+        let mut chunks = vec![MultiProofTargets::default(); 1];
+
+        let accounts_chunk =
+            self.proof_targets.drain(..self.accounts_chunk_size.min(self.proof_targets.len()));
+        for (address, storage_slots) in accounts_chunk {
+            let storage_chunks =
+                storage_slots.into_iter().sorted().chunks(self.storages_chunk_size);
+
+            // Initialize the chunk with an address with no storage slots. We need to do this,
+            // because the account may have no storage slots in the targets, but we still need to
+            // add it to the chunk.
+            chunks[0].entry(address).or_default();
+
+            for (i, chunk) in storage_chunks.into_iter().enumerate() {
+                if i >= chunks.len() {
```

Comment:
this looks a bit strange, not obvious what this does in detail only that this does a lot of alloc and sorting

EDIT: I think I understand what this does now

Reply:
how would you improve the docs?